stubs/SQLAlchemy/sqlalchemy/testing/provision.pyi

```diff
@@ -1,33 +1,117 @@
-from typing import Any
+from _typeshed import Incomplete, Self
+from collections.abc import Callable
+from logging import Logger
+from typing import Generic, Protocol, TypeVar
 
-log: Any
-FOLLOWER_IDENT: Any
+from sqlalchemy.engine.url import URL
+from sqlalchemy.testing.config import Config
 
-class register:
-    fns: Any
-    @classmethod
-    def init(cls, fn): ...
-    def for_db(self, *dbnames): ...
-    def __call__(self, cfg, *arg): ...
+log: Logger
+FOLLOWER_IDENT: Incomplete | None
 
 def create_follower_db(follower_ident) -> None: ...
-def setup_config(db_url, options, file_config, follower_ident): ...
+def setup_config(db_url, options, file_config, follower_ident) -> Config: ...
 def drop_follower_db(follower_ident) -> None: ...
 def generate_db_urls(db_urls, extra_drivers) -> None: ...
-def generate_driver_url(url, driver, query_str): ...
-def drop_all_schema_objects_pre_tables(cfg, eng) -> None: ...
-def drop_all_schema_objects_post_tables(cfg, eng) -> None: ...
 def drop_all_schema_objects(cfg, eng) -> None: ...
-def create_db(cfg, eng, ident) -> None: ...
-def drop_db(cfg, eng, ident) -> None: ...
-def update_db_opts(cfg, db_opts) -> None: ...
-def post_configure_engine(url, engine, follower_ident) -> None: ...
-def follower_url_from_main(url, ident): ...
-def configure_follower(cfg, ident) -> None: ...
-def run_reap_dbs(url, ident) -> None: ...
 def reap_dbs(idents_file) -> None: ...
-def temp_table_keyword_args(cfg, eng) -> None: ...
-def prepare_for_drop_tables(config, connection) -> None: ...
-def stop_test_class_outside_fixtures(config, db, testcls) -> None: ...
-def get_temp_table_name(cfg, eng, base_name): ...
-def set_default_schema_on_connection(cfg, dbapi_connection, schema_name) -> None: ...
+
+###
+# The methods below are decorated with the method register.init
+# They dynamically become instances of register,
+# wich is callable with their original parameters.
+#
+# While both mypy and pyright's validation work, Pylance is unable to
+# show the parameters and return types.
+#
+# There is a workaround (define the method for Pylance, then reassign
+# an instance of register to it for mypy, and add pyright+Flake8
+# suppressions), but it is too hacky and relies on some unsupported quirks.
+#
+# Asking Pylance to add support for these generic callables might be preferable.
+###
+
+_F = TypeVar("_F", bound=Callable[..., str | URL | None])
+_S = TypeVar("_S", bound=str)
+
+class register(Generic[_F]):
+    fns: dict[str, _F]
+    @classmethod
+    def init(cls: type[Self], fn: _F) -> Self: ...
+    def for_db(self: Self, *dbnames: str) -> Callable[[_F], Self]: ...
+    __call__: _F
```

Comment:
Same here -- maybe add a test case or two?

Reply:
It looks like the tests I can do for `register` are really limited. For example, because it's generic, I get `[assert-type]` errors like:
`Expression is of type "register[Callable[[object, object, object], None]]", not
"register[Any]"`, not "register[Any]"` or `Expression is of type "register[Callable[[object, object, object], None]]", not
"register[Callable[[object, object, object], None]]"`

At least I can test that the first parameter is correctly replaced.

https://github.com/python/typeshed/actions/runs/4045446150/jobs/6956947952

Reply:
(tests added in a previous commit)